cmd/kubeadm/app/cmd/join.go

```diff
@@ -419,6 +421,19 @@ func (j *joinData) Run() error {
 		fmt.Printf("[join] Running pre-flight checks before initializing the new control plane instance\n")
 		preflight.RunInitMasterChecks(utilsexec.New(), initCfg, j.ignorePreflightErrors)
 
+		fmt.Printf("[join] Pulling control-plane images")
+		containerRuntime, err := utilruntime.NewContainerRuntime(utilsexec.New(), j.cfg.GetCRISocket())
+		// don't fail, instead print a warning so the node can join the master, without pulling images.
+		if err != nil {
+			fmt.Printf("[join] failed to init container runtime, please retry with 'kubeadm config images pull': err: %s", err)
+		}
+
+		imagesPull := NewImagesPull(containerRuntime, images.GetAllImages(&initCfg.ClusterConfiguration))
+		// if the control-plane has no network, print a warning so user can pull image later
+		if err := imagesPull.PullAll(); err != nil {
+			fmt.Printf("[join] failed to pull control-plane images, please retry with 'kubeadm config images pull': err: %s", err)
```

Comment:
this approach seems OK to me.
please change the line to:
```
klog.Warningf("[join] WARNING: failed to pull images: %v. You can pre-pull images using the following command 'kubeadm config images pull', err)
```

/assign @rosti @bart0sh 
for comments on this warn-but-not-fail approach.

Reply:
actually, we might just error out here, because this is not an error that we can ignore.

please change the line to:
```
return errors.Wrapf(err, "failed to pull images. You can pre-pull images using the following command 'kubeadm config images pull'. error")
```

Reply:
Do we want to error out always? For example, if we are working with a pre-release of Kubernetes  with kubeadm (as when testing this PR) I actually got this error because it couldn't pull `v1.14` (despite the images of course existed locally). This might be an issue more with the environment I'm using to develop than anything else though, but aborting the execution here might not be the best option.

Reply:
Okay, I didn't check the internals of the pulling logic, just wanted to make sure that if the image exists we shouldn't error out even if there's no internet connection. I agree that if the image is missing then we should error out if the pulling fails.

Reply:
Looking at the code @neolit123 I think we just do a `docker pull <image>` with `runtime.exec.Command` and we'd be bubbling up that error, unless I'm missing something.

Reply:
erm, i think my previous comment still applies:

> hm, if the images exist locally the pull should be transparent and should not error out.
but if the images are missing and there is no internet for instance i think we should error out and instruct the user to pre-pull.

unless i'm missing something.

cc @rosti @bart0sh

Reply:
updated. Thx @neolit123 and @ereslibre for discussing this. updated

Reply:
The join process is aborted even if the image already existed:

```
    kubernetes_ha_master2: failed to pull images. You can pre-pull images using the following command 'kubeadm config images pull'. error: failed to pull image "k8s.gcr.io/kube-apiserver:v1.14.0": output: Error response from daemon: manifest for k8s.gcr.io/kube-apiserver:v1.14.0 not found
    kubernetes_ha_master2: , error: exit status 1
```

And, in that machine:

```
vagrant@master2:~$ docker images | grep k8s.gcr.io/kube-apiserver
k8s.gcr.io/kube-apiserver            v1.14.0                               594be9fb2b15        6 days ago          201MB
k8s.gcr.io/kube-apiserver            v1.14.0-alpha.0.1530_1b28775db1290a   594be9fb2b15        6 days ago          201MB
```

Reply:
yup, i have also reproduced the same behavior for this. I think waht @ereslibre  says make sense in this context. so i'm unresolving the conversation. :rocket:

Reply:
Oh @neolit123 maybe you meant if we used: https://github.com/kubernetes/kubernetes/blob/148248353b883f0dbbefa38100b675a7ed3ba587/cmd/kubeadm/app/preflight/checks.go#L829-L845 ?

Reply:
replied in a comment bellow about `RunPullImagesCheck`